spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -877,15 +911,32 @@ struct ProductionVisitor
             pushState(std::move(pstate));
         }
 
-        if ( auto a = AttributeSet::find(p.unitType().attributes(), "&size") ) {
+        // `&size` and `&max-size` shared the same underlying infrastructure
+        // so try to extract both of them and compute the ultimate value.
+        std::optional<Expression> size, maxSize, length;
+        if ( auto a = AttributeSet::find(p.unitType().attributes(), "&size") )
+            size = builder::coerceTo(*a->valueAs<Expression>(), type::UnsignedInteger(64));
+        if ( auto a = AttributeSet::find(p.unitType().attributes(), "&max-size") )
+            // If a `&max-size` was requested create a limited view one larger with
+            // the last byte acting as sentinel. Once parsing has finished we can
+            // then validate that parsing hasn't reached the sentinel byte.
+            maxSize = builder::incrementPrefix(builder::coerceTo(*a->valueAs<Expression>(), type::UnsignedInteger(64)));
+
+        if ( size && maxSize )
+            hilti::logger().internalError("&size and &max-size cannot be combined");
+        else if ( maxSize )
+            length = maxSize;
+        else if ( size )
+            length = size;
```

Comment:
This or the finalizer code currently triggers an HILTI error with the following unit decl,

```.ruby
type Foo = unit {
    xs: bytes &eod;
} &max-size=1;
```

We also should add tests for this (probably similar to the field-level tests).